Add a query check for unit tests for the `update_option()` function

mukeshpanchal27 opened the pull request:
This ticket will add query check for the `update_option()` function.

Trac ticket: https://core.trac.wordpress.org/ticket/22192

mukeshpanchal27 committed:
Add query check

felixarntz reviewed (commented):
@mukeshpanchal27 Looking at the failures (e.g. https://github.com/WordPress/wordpress-develop/actions/runs/6264894785/job/17012567207?pr=5272), it looks like in all cases there is 1 additional DB query happening. So we have to figure out where that comes from and then update the test accordingly.

I wonder if it somehow is a database query triggered by `get_option()`, which is called at the beginning of `update_option()`? Maybe you can log the database queries that occur in those tests to figure out what query is run?

Once we have figured that out, we may simply need to change the tests to expect 2 queries on `$update = true` and 1 query on `$update = false`.

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -381,13 +381,17 @@ public function data_option_autoloading() {
 	public function test_update_loosey_options( $old_value, $new_value, $update = false ) {
```
Let's take the 28 failures reported by this test.

Most if not all the failing datasets are triggering a query in `update_option()` via `$wpdb->update()`.

Note from the MySQL documentation for UPDATE:
> If you set a column to the value it currently has, MySQL notices this and does not update it.
> [Reference](https://dev.mysql.com/doc/refman/8.0/en/update.html#:~:text=If%20you%20set%20a%20column%20to%20the%20value%20it%20currently%20has%2C%20MySQL%20notices%20this%20and%20does%20not%20update%20it.)

This returns `0` for the number of affected rows and `$wpdb` [increments the value of `wpdb::$num_queries`](https://github.com/wordpress/wordpress-develop/blob/fe71ba1a1b5a833df25323e3a18fbdc4a35c588e/src/wp-includes/class-wpdb.php#L2337).

When `$wpdb->update()` returns `0` for the number of affected rows (via `$wpdb::query()`), `update_option()` [catches this in a falsy check](https://github.com/wordpress/wordpress-develop/blob/fe71ba1a1b5a833df25323e3a18fbdc4a35c588e/src/wp-includes/option.php#L824) and returns `false`.

This reveals the importance of the `get_num_queries()` tests here, as the `assertFalse()` in the tests was providing a false sense of confidence that a database query had not been performed.

Note the return annotation for `update_option()`:
```php
* @return bool True if the value was updated, false otherwise.
```
So when `update_option()` returns `false`, it tells us that the value was not changed. It doesn't tell us whether a query was run.

joemcgill replied:
This is a great observation, which wasn't clear by reading the docs for `$wpdb->update()`.

mukeshpanchal27 committed:
Add query checks

mukeshpanchal27 commented:
Thank you, @felixarntz and @costdev, for the feedback. I attempted to adjust the database query for the dataset, but it only seems to work for the `test_update_loosey_options` test. Unfortunately, other tests are failing consistently in the process.

felixarntz commented:
@mukeshpanchal27 I think the problem is that now you handle the number of queries in the data provider, but it needs to be handled in each of the three tests itself. The tests where the cache _not_ deleted are likely using an additional database request, while the ones with cache don't. Please try to resolve this at the individual test level instead of another parameter from the data provider which is used in all three tests.

Also, I believe in every test, whichever the correct number of database queries is, the number of queries in the `$update = true` cases should be 1 higher than the number of queries in the `$update = false` cases.

mukeshpanchal27 committed:
Simplify the tests.

Thanks @costdev for your help.

costdev commented:
@felixarntz I had a similar opinion as you, so I sent some revised tests to @mukeshpanchal27 that separate out the different cases (which he has committed above). This should hopefully provide a more well-rounded idea of how Core currently behaves when updating from/to certain values. If any additional cases come to mind, we can add these for additional coverage.

We can then see which tests fail with the new code, and evaluate which are expected/unexpected failures.

felixarntz commented:
@mukeshpanchal27 @costdev While those tests look better functionally, not sure I like that they are now split out.

More difficult to overview the different cases, and much harder to tell whether the tests were modified - the point of having them committed before was to _not_ have to modify them heavily. Adding the database query assertion was only about _adding_ something, but now the tests look completely different. Why can't we keep them together and keep the `$update` boolean parameter? I don't see a good reason that they were split out.

costdev commented:
@felixarntz I split out the tests because there are various additional queries going on in `trunk` even when the cache is being used. Determining the expected number of queries in one test method would add a lot of additional logic to be run for _each_ dataset. Splitting them up also makes it clearer which current WordPress behaviours seem "wrong" and may be improvable for performance gains.

Committing tests to trunk was to mainly to establish current behaviours I thought - And yes, ideally this would have also meant minimal modification to the tests. However, since we didn't test the number of queries, the current tests in trunk are somewhat incomplete and as shown by the number of new test methods needed to cover the number of queries, misleading. The updated tests are an effort to better cover the return value of `update_option()` _and_ number of queries it triggers.

FWIW: I'm not suggesting the new tests are written in the best possible way. If I saw a failure that indicated anomalous behaviour, I split it out to cover that behaviour. Having them split should hopefully make it easier for us to see where some might be merged into one test method, and where others need to remain standalone.

mukeshpanchal27 committed:
Merge branch 'WordPress:trunk' into test/22192-query

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
+		$num_queries = get_num_queries();
+		$updated     = update_option( 'foo', $new_value );
+
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'No additional queries should have run.' );
```
```suggestion
		$this->assertSame( 1, get_num_queries() - $num_queries, 'One additional query should have run.' );
```

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
-		} else {
-			$this->assertFalse( $updated, 'Loosely equal option should not trigger an update.' );
-		}
+		$this->assertSame( 2, get_num_queries() - $num_queries, 'One additional query should have run to update the value.' );
```
```suggestion
		$this->assertSame( 2, get_num_queries() - $num_queries, 'Two additional queries should have run.' );
```

costdev committed:
Fix `$message` argument.

costdev committed:
Fix `$message` argument.

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
+		$this->assertFalse( $updated, 'update_option() should have returned false.' );
+	}
+
+	/**
+	 * Tests that update_option() triggers two additional queries and returns false
+	 * for some loosely equal old and new values when the old value is retrieved from the database.
+	 *
+	 * The additional queries are triggered to:
+	 * - retrieve the old value from the database, as the option does not exist in the cache.
+	 * - attempt to update the value
+	 *
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_strictly_equal_values_that_should_not_update
+	 *
+	 * @param mixed $old_value The old value.
+	 * @param mixed $new_value The new value to try to set.
+	 */
+	public function test_update_option_should_trigger_two_queries_and_return_false_for_some_strictly_equal_values_from_db( $old_value, $new_value ) {
+		global $wpdb;
+
+		add_option( 'foo', $old_value );
+
+		// Delete cache.
+		wp_cache_delete( 'alloptions', 'options' );
+
+		$num_queries = get_num_queries();
+
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( 2, get_num_queries() - $num_queries, 'Two additional queries should have run to retrieve the old value from the database.' );
+		$this->assertSame( 0, $wpdb->rows_affected, 'No rows should have been affected as the database considered the values to be the same.' );
+		$this->assertFalse( $updated, 'update_option() should have returned false.' );
+	}
```
```suggestion
	/**
	 * Tests that update_option() triggers one additional query and returns false
	 * for some loosely equal old and new values when the old value is retrieved from the database.
	 *
	 * The additional query is triggered to retrieve the old value from the database,
	 * as the option does not exist in the cache.
	 *
	 * @ticket 22192
	 *
	 * @covers ::update_option
	 *
	 * @dataProvider data_update_option_strictly_equal_values_that_should_not_update
	 *
	 * @param mixed $old_value The old value.
	 * @param mixed $new_value The new value to try to set.
	 */
	public function test_update_option_should_trigger_one_additional_query_and_return_false_for_some_strictly_equal_values_from_db( $old_value, $new_value ) {
		global $wpdb;

		add_option( 'foo', $old_value );

		// Delete cache.
		wp_cache_delete( 'alloptions', 'options' );

		$num_queries = get_num_queries();

		$updated = update_option( 'foo', $new_value );

		$this->assertSame( 1, get_num_queries() - $num_queries, 'One additional query should have run to retrieve the old value from the database.' );
		$this->assertSame( 0, $wpdb->rows_affected, 'No rows should have been affected as the database considered the values to be the same.' );
		$this->assertFalse( $updated, 'update_option() should have returned false.' );
	}
```

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
+
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'One additional query should have run.' );
+		$this->assertFalse( $updated, 'update_option() should have returned false.' );
+	}
```
```suggestion
	/**
	 * Tests that update_option() triggers no additional queries and returns false
	 * for some strictly equal old and new values when the cache is refreshed.
	 *
	 *
	 * @ticket 22192
	 *
	 * @covers ::update_option
	 *
	 * @dataProvider data_update_option_strictly_equal_values_that_should_not_update
	 *
	 * @param mixed $old_value The old value.
	 * @param mixed $new_value The new value to try to set.
	 */
	public function test_update_option_should_trigger_no_additional_queries_and_return_false_for_some_strictly_equal_values_from_refreshed_cache( $old_value, $new_value ) {
		add_option( 'foo', $old_value );

		// Delete and refresh cache from DB.
		wp_cache_delete( 'alloptions', 'options' );
		wp_load_alloptions();

		$num_queries = get_num_queries();
		$updated     = update_option( 'foo', $new_value );

		$this->assertSame( $num_queries, get_num_queries(), 'No additional queries should have run.' );
		$this->assertFalse( $updated, 'update_option() should have returned false.' );
	}
```

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
+		);
+	}
+
+	/**
+	 * Tests that update_option() triggers one additional query and returns false
+	 * when there are unaffected database rows.
+	 *
+	 * This excludes old values that are null, as they trigger yet another additional query.
+	 *
+	 * The additional query is triggered to perform an UPDATE despite the values being loosely equal.
+	 *
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_loosely_equal_values_that_should_trigger_one_query_and_return_false
+	 *
+	 * @param mixed $old_value The old value.
+	 * @param mixed $new_value The new value to try to set.
+	 */
+	public function test_update_option_should_trigger_one_query_and_return_false_for_some_loosely_equal_values( $old_value, $new_value ) {
+		global $wpdb;
+
+		add_option( 'foo', $old_value );
+
+		$num_queries = get_num_queries();
+
+		// Comparison will happen against value cached during add_option() above.
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( 1, get_num_queries() - $num_queries, 'One additional query should have run to update the value.' );
+		$this->assertSame( 0, $wpdb->rows_affected, 'No rows should have been affected as the database considered the values to be the same.' );
+		$this->assertFalse( $updated, 'update_option() should have returned false.' );
+	}
```
```suggestion
	/**
	 * Tests that update_option() triggers no additional queries and returns false
	 * when there are unaffected database rows.
	 *
	 * This excludes old values that are null, as they trigger an additional query.
	 *
	 * @ticket 22192
	 *
	 * @covers ::update_option
	 *
	 * @dataProvider data_update_option_loosely_equal_values_that_should_trigger_one_query_and_return_false
	 *
	 * @param mixed $old_value The old value.
	 * @param mixed $new_value The new value to try to set.
	 */
	public function test_update_option_should_trigger_no_queries_and_return_false_for_some_loosely_equal_values( $old_value, $new_value ) {
		global $wpdb;

		add_option( 'foo', $old_value );

		$num_queries = get_num_queries();

		// Comparison will happen against value cached during add_option() above.
		$updated = update_option( 'foo', $new_value );

		$this->assertSame( $num_queries, get_num_queries(), 'No additional queries should have run.' );
		$this->assertSame( 0, $wpdb->rows_affected, 'No rows should have been affected as the database considered the values to be the same.' );
		$this->assertFalse( $updated, 'update_option() should have returned false.' );
	}
```

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,437 @@ public function data_option_autoloading() {
+		);
+	}
+
+	/**
+	 * Tests that update_option() triggers two additional queries and returns false
+	 * when the old value was null and the new value is loosely equal.
+	 *
+	 * @ticket 22192
+	 *
+	 * @covers ::update_option
+	 *
+	 * @dataProvider data_update_option_should_trigger_two_queries_and_return_false_for_a_null_old_value_and_some_loosely_equal_new_values
+	 *
+	 * @param mixed $new_value The new value to try to set.
+	 */
+	public function test_update_option_should_trigger_two_queries_and_return_false_for_a_null_old_value_and_some_loosely_equal_new_values( $new_value ) {
+		add_option( 'foo', null );
+
+		$num_queries = get_num_queries();
+
+		// Comparison will happen against value cached during add_option() above.
+		$updated = update_option( 'foo', $new_value );
+
+		$this->assertSame( 2, get_num_queries() - $num_queries, 'Two additional queries should have run.' );
+		$this->assertFalse( $updated, 'update_option() should have returned false.' );
+	}
```
```suggestion
	/**
	 * Tests that update_option() triggers one additional queries and returns false
	 * when the old value was null and the new value is loosely equal.
	 *
	 * @ticket 22192
	 *
	 * @covers ::update_option
	 *
	 * @dataProvider data_update_option_should_trigger_two_queries_and_return_false_for_a_null_old_value_and_some_loosely_equal_new_values
	 *
	 * @param mixed $new_value The new value to try to set.
	 */
	public function test_update_option_should_trigger_one_additional_query_and_return_false_for_a_null_old_value_and_some_loosely_equal_new_values( $new_value ) {
		add_option( 'foo', null );

		$num_queries = get_num_queries();

		// Comparison will happen against value cached during add_option() above.
		$updated = update_option( 'foo', $new_value );

		$this->assertSame( 1, get_num_queries() - $num_queries, 'One additional query should have run.' );
		$this->assertFalse( $updated, 'update_option() should have returned false.' );
	}
```

mukeshpanchal27 committed:
Apply suggestions from code review

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

mukeshpanchal27 commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,157 +370,432 @@ public function data_option_autoloading() {
+	 * Tests that update_option() triggers no additional queries and returns false
+	 * for some strictly equal old and new values when the cache is refreshed.
+	 *
+	 *
```
```suggestion
```

mukeshpanchal27 committed:
Apply suggestions from code review

Commit:
Remove `rows_affected` assertions.

Previously, a query ran to try to update the value.
The values were considered equal, so this resulted in 0 rows affected.

Since the query no longer runs, there is no need to check the number
of rows affected.

costdev committed:
Resolve test failures and improve test organisation.

This resolves test failures by reducing each of the expected
queries by 1, which is the expected behaviour after this patch.

The tests have been restructured to minimize the number of
test methods while ensuring that happy paths and unhappy
paths remain separate.

mukeshpanchal27 commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,193 +370,329 @@ public function data_option_autoloading() {
+	 */
+	public function data_stored_as_empty_string() {
+		return array(
+			'false'        => array( false ),
```
```suggestion
```

mukeshpanchal27 commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,193 +370,329 @@ public function data_option_autoloading() {
+			'empty string'       => array( '', '' ),
+			'(int) 0'            => array( 0, 0 ),
+			'(float) 0.0'        => array( 0.0, 0.0 ),
+			'empty array'        => array( array(), array() ),
```
```suggestion
			'empty array'        => array( array(), array() ),
			'false'              => array( false, false ),
```

mukeshpanchal27 committed:
Apply suggestions from code review

mukeshpanchal27 committed:
Update code for BC

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -569,20 +705,22 @@ static function () use ( $default_value ) {
 		);
 
-		update_option( $option, $default_value );
+		$this->assertTrue( update_option( $option, $default_value ), 'update_option() should have returned true.' );
```
In the old `update_option()`, if the old value and new value were strictly equal, it would return `false`, whether or not the old value matched the default. [See here](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/option.php#L501).

In this test:
- The option doesn't exist yet.
- The default value has been filtered to `default-value`.
- The new value is `default-value`.

So this test will actually return `false`.

Since this is supposed to be testing that a non-existent option with a filtered default value is added to the database, we need to make a couple of changes so that the new value isn't the same as the default value.

Change 1:
```suggestion
		$this->assertTrue( update_option( $option, 'new value' ), 'update_option() should have returned true.' );
```

costdev replied:
Unresolving so it's available for reference.

costdev commented on tests/phpunit/tests/option/option.php:
```diff
@@ -569,20 +705,22 @@ static function () use ( $default_value ) {
-		$this->assertSame( 1, $actual );
+		$this->assertIsObject( $actual, 'The option was not added to the database.' );
+		$this->assertObjectHasProperty( 'option_value', $actual, 'The "option_value" property was not included.' );
+		$this->assertSame( $default_value, $actual->option_value, 'The value was not stored as an empty string.' );
```
Change 2:
```suggestion
		$this->assertSame( 'new value', $actual->option_value, 'The new value was not stored in the database.' );
```

mukeshpanchal27 committed:
Apply suggestions from code review

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

costdev commented:
@joemcgill @felixarntz While investigating test failures, myself and @mukeshpanchal27 discovered a BC break in the new `update_option()`.

For a non-existent option, the old `update_option()` would return `false` if the old and new value were strictly equal ([Reference](https://github.com/WordPress/wordpress-develop/blob/6.2/src/wp-includes/option.php#L501)). This occurred even if the old value matched the default value.

In the new `update_option()`, we were adding the option even if the old and new values were strictly equal -> BC break.

To verify this, we committed [1747534](https://github.com/WordPress/wordpress-develop/pull/5272/commits/17475343937d243be4d14d563dd2b1bc6ae289af). As expected, this failed because a query was performed via `add_option()`, which did not occur in the old `update_option()`.

To resolve this, we committed [cc751bf](https://github.com/WordPress/wordpress-develop/pull/5272/commits/cc751bfada93bfea1de51d39aaee57f49d7b790a).

In addition, the `test_update_option_should_add_option_with_filtered_default_value` method required some changes. Details can be found [here](https://github.com/WordPress/wordpress-develop/pull/5272#discussion_r1340914515), and the changes were committed in [9a91d74](https://github.com/WordPress/wordpress-develop/pull/5272/commits/9a91d74823e2968fba822a2ef0d03505070c6b7f).

Finally, @mukeshpanchal27 manually tested the [timezone issue](https://core.trac.wordpress.org/ticket/22192#comment:52) to verify that these commits did not cause the regression to occur again.

joemcgill commented:
@mukeshpanchal27 and @costdev, thanks for digging into this and for providing such a thorough explanation of the issues you've discovered while investigating the unexpected behavior with the initial set of query tests that were added in [fe71ba1](https://github.com/WordPress/wordpress-develop/pull/5272/commits/fe71ba1a1b5a833df25323e3a18fbdc4a35c588e). 

There's a lot to unpack in this thread, but to summarize—for each use case we now need to test for three distinct things:

1. **Does `update_option` result in `true` or `false`?** This is what our initial set of test cases were put in place to check.
2. **Is an additional DB request being made when `update_option` is `false`?** Our hypothesis, which has been proven false, is that no additional DB requests would be made if the option was not being updated.
3. **Should any additional DB request be expected, or are they a bug (i.e., should any of these queries be avoided)?** More on this below...

The cases where we're confirming that `update_option()` returns false, but resulting in an additional DB query anyway, seems like the place where we _might_ want to add logic to avoid those queries. However, reviewing all of the changes that have been made in this PR, you've updated the test cases AND the logic, which makes it really difficult to confirm what is being changed and why. So I am agreement with what @felixarntz mentioned in his [previous comment](https://github.com/WordPress/wordpress-develop/pull/5272#issuecomment-1733880963).

Looking back at the original commit that added query count checks, I noticed that there were some flaws in the tests themselves that were contributing to the high number of the failures. I've created [this PR](https://github.com/WordPress/wordpress-develop/pull/5356) which goes back to the original tests that @mukeshpanchal27 added, but then fixes the way query counts were being applied, and am only seeing 9 unexpected failures—all of which have to do with the unique behavior you observed for when the initial option is set to either `false` or `null`.

I'd be much more confident at this point in adjusting the `test_update_loosey_options()` test method to account for these additional queries, rather than refactoring all of the original test cases we added. I think it leads to much more confidence in the changes we're making and avoiding unintentional regressions.

costdev commented:
Thanks @joemcgill!

> Looking back at the original commit that added query count checks, I noticed that there were some flaws in the tests themselves that were contributing to the high number of the failures.

Can you elaborate on this?

costdev commented:
I think part of the issue is that the tests currently in trunk don't check for the number of expected queries, but the logic changes to `update_option()` are also now in trunk.

It may be prudent to add the query count checks to the tests in trunk first, so that any changes this PR makes are only to _adjust_ the query count checks, rather than to introduce them.

However, the query count checks need to be based on trunk _before_ the logic changes to `update_option()` were committed (i.e. 6.3's `update_option()`). Otherwise, we're protecting the new logic, not BC.

I'm finding it difficult to write the query count checks for trunk's tests as so much is covered in one test method, while there's actually more going on under the hood in `update_option()` - hence splitting up the tests. Given this, @joemcgill would you be willing to write the checks for query counts for the tests in trunk, based on 6.3's `update_option()`?

Also, since tests to protect BC will actually fail in trunk because the new logic for `update_option()` is already committed, we may need to temporarily revert the new logic so accurate tests can be added first. Then we can commit the new logic along with _adjusted_ tests.

mukeshpanchal27 commented:
@joemcgill Thanks for your review. I have open [PR 5362](https://github.com/WordPress/wordpress-develop/pull/5362) with the query check that you did in your [PR 5356](https://github.com/WordPress/wordpress-develop/pull/5356) it result in [78 failures in query checks](https://github.com/WordPress/wordpress-develop/actions/runs/6376169336/job/17302847099?pr=5362).

spacedmonkey commented on src/wp-includes/option.php:
```diff
@@ -805,7 +805,13 @@ function update_option( $option, $value, $autoload = null ) {
 	 */
-	if ( $raw_old_value !== $default_value && _is_equal_database_value( $raw_old_value, $value ) ) {
+	if (
+		$value === $raw_old_value ||
```
I am not sure why this is needed, isnt this handled by `_is_equal_database_value`?

mukeshpanchal27 replied:
We got one BC issue [raise here](https://core.trac.wordpress.org/ticket/22192#comment:52) and that we addressed in [changeset 56717](https://core.trac.wordpress.org/changeset/56717). Hope that commit message answers your queries.

spacedmonkey replied:
Okay, what about 

```
_is_equal_database_value( $raw_old_value, $value ) || $raw_old_value !== $default_value
```

Is that the same?

https://github.com/WordPress/wordpress-develop/blob/8bbeee43b647d6c8fc54c67cb0616343ac7b03fb/src/wp-includes/option.php#L2934-L2936

costdev replied:
The bail cases are:
- The old and new values are strictly equal.
- The old and new values are loosely equal AND the old value is not the default value. - Note that this case has two conditions.

spacedmonkey replied:
Maybe we should rename `_is_equal_database_value` to say loosely equal, to make this easier to read then.

costdev replied:
Maybe something like `_database_converts_to_same_value()`?

NTIH.

felixarntz replied:
I am not convinced that this additional clause here is needed either.

I have read through https://github.com/WordPress/wordpress-develop/pull/5272#issuecomment-1740347838, and I acknowledge that technically storing the option if the given value is the same as the default is a backward compatibility break. However, we have also previously clarified that _not_ allowing that leads to another problem because the `get_option()` result of this scenario would in fact be different, so it _should_ set the value in the database:
* For a non-existent option, `get_option()` returns `false`.
* For an option stored with `update_option( 'foo', false )`, `get_option()` returns `''`.

See for example @joemcgill's comment in https://github.com/WordPress/wordpress-develop/pull/5139#discussion_r1323340693: The above example leads to a different return type value, so I think it should result in an update.

This is a crucial differentiation. Unless you have encountered this BC break in real-world examples, I personally think this BC break is an edge case that is okay to accept, for two reasons:
* I cannot imagine that someone calls `update_option( 'foo', false )` and explicitly wants that value to not be stored in the DB.
* Regarding the return value, the majority of plugins don't care about the return value of `update_option()` and assume it works as expected.

costdev replied:
@felixarntz Since this ticket is about reducing queries by comparing values, I don't think creating a new option in the database when `old value === new value === default value` is within that scope, and such a change would likely need its own commit for better tracking of history.

This strict check prevents an unnecessarily and unplanned regression/change to `update_option()`'s behaviour.

felixarntz replied:
@costdev I _may_ agree with that.

Main point though is that we should decouple the production code change here from the tests added, since by far most tests here were just for improved coverage of the original change, so the production code fix here, whether it gets committed or not, should then also go in on its own IMO.

joemcgill commented:
@costdev: 

> Can you elaborate on this?

Sure! The original query count checks that were added in fe71ba1 didn't account for the fact that when the options cache is deleted, that `update_option()` will always run a query to get the previous value before performing the update. I adjusted the logic in those tests in https://github.com/WordPress/wordpress-develop/pull/5356/commits/5aa21d5f36b1c1c6f7e1354efbeb9321a5808e2a in my PR to account for this expected behavior.

> I think part of the issue is that the tests currently in trunk don't check for the number of expected queries, but the logic changes to `update_option()` are also now in trunk.

That's true. To see what the previous behavior was prior to the changes committed in [r56681](https://core.trac.wordpress.org/changeset/56681 ) and [r56717](https://core.trac.wordpress.org/changeset/56717), I branched from when the initial tests were added (6b50b1b) and then applied the unit query count tests (cherry picking 4521a53 and 03e0241) and am seeing nearly the same result that Mukesh demonstrated in https://github.com/WordPress/wordpress-develop/pull/5362. I've pushed that branch to my fork so you can see [the comparison](https://github.com/WordPress/wordpress-develop/compare/6b50b1b...joemcgill:wordpress-develop:update/22192-tests-prior-to-changes), but you'd have to run the tests locally to see the output, since the GH actions have to run against a specific branch of the repo and not against an arbitrary commit.

Prior to the changes we've committed, I'm seeing 74 test failures due to queries happening when we expect them not to, but now that those changes are in trunk, we're only seeing 9 failures. What I think that demonstrates, is that the changes we've committed have fixed a majority of the cases where extra queries were being made when they didn't need to be. I.e., this demonstrates that the bug is fixed in many cases, rather than showing a back-compat break, unless I'm missing something.

I think the next steps would be to adjust the PR that I started in https://github.com/WordPress/wordpress-develop/pull/5356 so that the 9 cases that are still failing are accounted for. This should happen by either:

1. Updating the logic in those tests so they expect the extra queries in these 9 cases
2. Fixing the update_option logic so those extra queries are no longer being made.

I've not looked deep enough into why those cases are failing to know whether those queries should be expected or if they can be avoided without additional side effects.

Hopefully this helps? Let me know if anything is unclear.

costdev commented:
Thanks @joemcgill! That helps!

-----

Just noting here as clarification on the BC break that was revealed while working on the tests: It's actually a regression rather than a BC break. My bad on that.

When running `update_option( false, false )`, for example:
- 6.3: Since both the old and new value are strictly equal, `update_option()` returned `false` without running an additional query.
- trunk: An additional query runs to try to add the option.
  - As the old and new value are considered the same by the database, this results in 0 affected rows, which `add_option()` catches and returns `false`. In turn, `update_option()` returns `false` too.
  - The original tests passed because they only checked `assertFalse( update_option( false, false ) )`, so they didn't reveal the additional query that previously didn't happen. Since adding test coverage for this additional query would be protecting the regression, we instead resolved it with [this commit](https://github.com/WordPress/wordpress-develop/pull/5272/commits/cc751bfada93bfea1de51d39aaee57f49d7b790a).

-----

> I think the next steps would be to adjust the PR that I started in https://github.com/WordPress/wordpress-develop/pull/5356 so that the 9 cases that are still failing are accounted for. This should happen by either:
> 
> - Updating the logic in those tests so they expect the extra queries in these 9 cases
> - Fixing the update_option logic so those extra queries are no longer being made.
>
> I've not looked deep enough into why those cases are failing to know whether those queries should be expected or if they can be avoided without additional side effects.

Resolving those 9 cases using that single test method may be tricky/awkward/messy.

Though to start off, I'd suggest adding [this commit](https://github.com/WordPress/wordpress-develop/pull/5272/commits/cc751bfada93bfea1de51d39aaee57f49d7b790a) to PR #5356 which resolves the regression. You'll see that the failure count increases from 9 to 10, but with some different tests failing. `test_update_option_should_add_option_with_filtered_default_value`, for example, was protecting the regression by expecting the option to be added despite strictly equal old and new values. This should be adjusted to pass the "default" default value of `false` to `update_option()`.

Overall though, I do feel that we've become too reliant on the incomplete tests we have in trunk, which asserted based on an incorrect association regarding `update_option()`'s return value and the query count. By continuing with that one test method, I feel we'd be trying to do too much in one test method just to recover from that mistake. While I'll go with consensus here, I'd much prefer if you could split up the tests from #5356 to `update_option_should_update()`, `update_option_should_not_update()`, `update_option_should_add_option()`, etc. and split the data provider up, ensuring that all of #5356's test cases are covered. We're testing multiple behaviours with one test method, which feels wrong and problematic for clarity and maintenance. It was different when we were in a rush towards Beta 1 and thought the tests were sound, but now that we know they're not, there's time to get the tests in better shape.

joemcgill commented:
Thanks @costdev. I think I'm on the same page as you now. 

I can see from @mukeshpanchal27's  PR that an extra query was being made in 6.3 when updating from `false` to `false` when updating without a primed cache ([example](https://github.com/WordPress/wordpress-develop/actions/runs/6376169336/job/17302848488#step:14:1065)) and also when the cache has been refreshed from the DB ([example](https://github.com/WordPress/wordpress-develop/actions/runs/6376169336/job/17302848488#step:14:1249)). Now in trunk, those extra queries are not being made, but an extra query _is_ being made if the option is already in cache from the first time it's set ([example](https://github.com/WordPress/wordpress-develop/actions/runs/6357048522/job/17267646802#step:14:727)). Addressing this regression via [this commit](https://github.com/WordPress/wordpress-develop/pull/5272/commits/cc751bfada93bfea1de51d39aaee57f49d7b790a) seems sensible to me. I don't see any other new failures as a result of the changes in trunk when cross referencing [this PR](https://github.com/WordPress/wordpress-develop/pull/5356) with the ones @mukeshpanchal27 added in https://github.com/WordPress/wordpress-develop/pull/5362

As far as splitting out the unit tests are concerned, I agree that we're making each test do too much and would be happy with a refactor as long as we're only changing tests and not tests AND functionality in the same PR, because then it's unclear how the logic change relates to the behavior that was previously being confirmed by the tests—which was the main concern I had in [this comment](https://github.com/WordPress/wordpress-develop/pull/5272#issuecomment-1741519838).

Personally, I'd rather separate the unit tests so that we're not mixing the `update_option` return value and the query counts logic in the same assertion, rather than separating tests and data providers based on whether or not they produce an update, but I don't feel strongly about which strategy is used.

Seems like the most important things to address are removing the extra query you identified and fixing the new tests that were added for filtered default values that were based on this regression. Agreed?

felixarntz commented on tests/phpunit/tests/option/option.php:
```diff
@@ -370,193 +370,329 @@ public function data_option_autoloading() {
+	public function data_stored_as_empty_string() {
+		return array(
+			'empty string' => array( '' ),
+			'null'         => array( null ),
```
Per the above, I think we should also include `false` here. In the database, `false` becomes `''`, so it is not the same as the default value `false`.

felixarntz reviewed (commented):
In terms of the additional tests for the original commit, I think this looks good to commit. However, I am not sure the single production code change here makes sense.

mukeshpanchal27 committed:
Revert production code changes

mukeshpanchal27 committed:
Revert unit tests chnages

mukeshpanchal27 committed:
Revert some more unit tests changes

felixarntz commented:
Committed in https://core.trac.wordpress.org/changeset/56762

@costdev @mukeshpanchal27 So this is now unblocking a follow up PR to further consider the production code change.
